### PR TITLE
CMake Update

### PR DESCRIPTION
**Pull request by phlptp**
Change the CMake to use subdirectory and move the library generation …and warnings to separate files.

### CODE REVIEW

**Comment by phlptp** [comment]
pre-commit.ci autofix

File: azure-pipelines.yml [hunk 1]
```diff
@@ -109,10 +109,6 @@ jobs:
         gcc8:
           containerImage: gcc:8
           cli11.std: 17
-        gcc4.8:
-          containerImage: gcc:4.8
```
Comment (henryiii) [inline azure-pipelines.yml:113]: Is there a reason this can't just change to `helics/buildenv:gcc4-8-builder`? Does `.ci/azure-cmake.yml` break?

Comment (phlptp) [reply azure-pipelines.yml:113]: My guess would be no, but CMake is already installed so just seemed a waste of time, I will put it back to make sure.

File: examples/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 function(add_cli_exe T)
-  add_executable(${T} ${ARGN} ${CLI11_headers})
+  add_executable(${T} ${ARGN})
```
Comment (henryiii) [inline examples/CMakeLists.txt:2]: By removing the headers, doesn't this mean you can't access these in an IDE anymore? (Though if you are using the build-ahead mode, you can - so maybe that's why?)

**Review by henryiii — APPROVED** [review]

File: examples/CMakeLists.txt (hunk 2 above)
Comment (phlptp) [reply examples/CMakeLists.txt:2]: In an IDE, the files are now added to the CLI11 Project, which is new.  They are added as build in precompile mode or with newer CMakes they are added as private sources in the interface library.

File: azure-pipelines.yml (hunk 1 above)
Comment (phlptp) [reply azure-pipelines.yml:113]: I put it back in the Docker section,  installing CMake takes about 30 seconds on the CI test, so we can reduce the build time slightly by leaving it in its own section, or just leave the extra CMake install in place.

Comment (henryiii) [reply azure-pipelines.yml:113]: We've got enough jobs where I'm not too worried about it. I'd rather like to move to GHA eventually & then we can probably use the CMake action. Keeping this simple makes it easier to move. ;)

**Pull request merged**